### ISSUE TITLE
Program failed to find the sgoinfre dir ( 42 Québec )

### ISSUE THREAD

**Issue by loyc12** [itislu/42free#22]
The program cannot locate the campus's sgoinfre directory, which is located at `computer/Macintosh_SSD/Sytem/Volumes/Macintosh_SSD/sgoinfre` ( + `/Perso/$USER` for our personal sgoinfre )

For reference, our home is located directly on `computer/home_$USER`, and contains a direct shortcut to sgoinfre as well

It's likely our campus' architecture or our usage of MacOS is to blame here

**Comment by itislu** [comment]
Thanks a lot for the report!

So if I understand correctly, the path to your personal sgoinfre does NOT contain your username?

**Comment by loyc12** [comment]
it would be the later, yes. altho it is also accessible through `computer/home_$USER/sgoinfre/Perso/$USER` due to the shortcut.

maybe letting users manually input their personal sgoinfre directory in case of such failure could be a good idea btw(?)

**Comment by itislu** [comment]
If it cannot be found, trying to find a shortcut from the home directory as a fallback could work.
At the moment this is the command how it tries to find the sgoinfre directory: 
```bash
find /sgoinfre/ -type d -name "$USER"
```
So atm it assumes the parent directory of all personal sgoinfre directories is in `/sgoinfre`.

I will try to make the search for it more flexible.

If it still cannot find one, prompting the user for manual input is a great idea! Thank you!

**Comment by loyc12** [comment]
running `find $HOME/**/sgoinfre/** -type d -name "$USER"` seems to find it, along with a lithany of "permission denied". however, I don't know if its compatible with your architecture

**Comment by itislu** [comment]
Oh that's great!
The "permission denied" errors will be silenced with `2>/dev/null`, no worries.

The only issue is that the real path of where the symbolic link links to will be needed.
- Usually it's possible to get that with the `realpath` command, but it sadly doesn't exist on macOS (#21).

**Comment by loyc12** [comment]
That is unfortunate. I guess you could create (steal) your own realpath implementation and ship it with the program, but that's seems clunky. a "specify your sgoinfre" failsafe seems simpler and more future-proof (?)

I tried running `find **/sgoinfre/** -type d -name "$USER"` and here's what it found after 1 minute
```
/System/Volumes/Data/Users/llord/sgoinfre/Perso/llord
/System/Volumes/Data/sgoinfre/Perso/llord
/Users/llord/sgoinfre/Perso/llord
```

**Comment by itislu** [comment]
Oh yes, in order to reduce the time it takes, the full line that 42free uses is the following:
```bash
find /sgoinfre/ -type d -name "$USER" -print -quit 2>/dev/null | grep -oE "^.*$USER" | head -n 1
```
But I think this one will not return the correct result for your campus' layout because the username appears twice in the path. The command above will just cut everything after the first appearance of the username.

For the three directories that you found, could try to `cd` into each of them and do `pwd -P` to double check that they're the real paths?

I will definitely implement a failsafe.

**Comment by loyc12** [comment]
Only the second one is real, as the first and last use the llord->sgoinfre shortcut.
Checking that the path uses $USER only once could garantee a correct (real) path, and it seems simple engouh to do. 
However, it's still a long command to run, so it needs optimizing imho.

Btw, I used the command `find /**/sgoinfre/** -type d -name "$USER" 2>/dev/null` and still got some zsh errors (probably de to my cmi) ;

```
➜  llord find /**/sgoinfre/** -type d -name "$USER" 2>/dev/null
zsh: permission denied: sharedVault
zsh: permission denied: com.apple.LaunchServices-1082-v2.csstore
zsh: permission denied: sharedVault
zsh: permission denied: com.apple.LaunchServices-1082-v2.csstore
/System/Volumes/Data/Users/llord/sgoinfre/Perso/llord
/System/Volumes/Data/sgoinfre/Perso/llord
/Users/llord/sgoinfre/Perso/llord
➜  llord
```

**Comment by JonatV** [comment]
You make me realize this access : `/sgoinfre/students/jveirman` in 19 brussels is no more than a symlink.
And it also located in the `/System/Volumes/Data/sgoinfre/students/jveirman`
There's a possibility the sgoinfre will always be located there in every school.

1. So when starting the script it can first check the most probable location by doing this :
`find /System/Volumes/Data/sgoinfre/* -type d -name "$USER" -print -quit  2>/dev/null  | grep -oE "^.*$USER" | head -n 1`

2. But if this cmd doesn't find any relevant repo : 
2.1. it ask the user to enter the path
2.2 Or we prevent the user that we would like to run a bigger script and it might take up to 1 minute. If yes we run the same `find` command but on the all system

**Comment by itislu** [comment]
@loyc12 @JonatV Could you try if this command works:
```bash
find / -path $HOME -prune -o -path '*/sgoinfre/*' -type d -name "$USER" -print -quit 2>/dev/null | grep -oE "^.*$USER" | head -n 1
```
This excludes the $HOME directory from the search, which should help with performance and avoid reporting any wrong directories also named sgoinfre by the user.

**Comment by JonatV** [comment]
> [@loyc12](https://github.com/loyc12) [@JonatV](https://github.com/JonatV) Could you try if this command works:
> 
> find / -path $HOME -prune -o -path '*/sgoinfre/*' -type d -name "$USER" -print -quit 2>/dev/null | grep -oE "^.*$USER" | head -n 1
> 
> This excludes the $HOME directory from the search, which should help with performance and avoid reporting any wrong directories also named sgoinfre by the user.
> 
> **UPDATE** loyc12 reported that sadly it cannot find the sgoinfre directory :/

It does work here, takes around 30 sec

**Comment by itislu** [comment]
@loyc12 @JonatV If you copy-paste the following script into your terminal, how long does it take to find your sgoinfre directory?
```bash
#!/bin/bash

# Breadth-first search to find a directory with a specific name and containing a specific pattern in its path
find_dir()
{
    local start_dir=$1
    local exclude_dir=$2
    local target_dir=$3
    local path_pattern=$4

    python3 -c "
import os
import queue

def bfs_find(start_dir, exclude_dir, target_dir, path_pattern):
    q = queue.Queue()
    q.put(start_dir)

    while not q.empty():
        current_dir = q.get()
        try:
            # Skip processing for the excluded directory
            if current_dir.startswith(exclude_dir):
                continue

            with os.scandir(current_dir) as it:
                for entry in it:
                    if entry.is_dir():
                        full_path = entry.path
                        dir_name = os.path.basename(full_path)

                        # Check if the directory name is the target and the path contains the pattern
                        if dir_name == target_dir and path_pattern in full_path:
                            print(full_path)
                            return  # Stop after the first match

                        # Add subdirectories to the queue for further processing
                        q.put(full_path)

        except PermissionError:
            # Handle directories for which the user does not have permissions
            continue

bfs_find('$start_dir', '$exclude_dir', '$target_dir', '$path_pattern')
"
}

find_dir "/" "$HOME" "$USER" "sgoinfre/"

```

**Comment by itislu** [comment]
Alternatively, this command would also be very interesting:
```bash
find / -maxdepth 12 -path "$HOME" -prune -o -path '*/sgoinfre/*' -type d -name "$USER" -print -quit | perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/"
```

**Comment by JonatV** [comment]
> @loyc12 @JonatV If you copy-paste the following script into your terminal, how long does it take to find your sgoinfre directory?
> 
> ```shell
> #!/bin/bash
> ...
> ```

Took 0.073 seconds

--------------------------------------

While :
>`find / -maxdepth 12 -path "$HOME" -prune -o -path '*/sgoinfre/*' -type d -name "$USER" -print -quit | perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/"`

Took 34.064 seconds

**Comment by loyc12** [comment]
the script itself gives me an error : 

```
Traceback (most recent call last):
  File "<string>", line 34, in <module>
  File "<string>", line 16, in bfs_find
OSError: [Errno 9] Bad file descriptor: '/dev/fd/3'
```

As for the command, it takes about 17 seconds, but gives me alot of permission errors, especially from :

```
/Library/
/System/Library/
/System/Volumes/Data/Library/
/System/Volumes/Data/System/Library/ ( duplicate of /System/Library/ )
/System/Volumes/Data/private/
```

**Comment by itislu** [comment]
@JonatV Thank you very much!

@loyc12 I updated the script and the command, could you try both again?

**Comment by itislu** [comment]
And could you also time this one please:
```bash
time find /sgoinfre /System/Volumes/Data/sgoinfre -maxdepth 6 -type d -name "$USER" -print -quit 2>/dev/null | perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/"
```
The goal would be that it this one is a quick first check, so it should not take longer than 1 second, ideally.

**Comment by loyc12** [comment]
@itislu I Tried the script and it works!!! I only had to add _bfs after the name of the function called ( you forgot to add it )

```
➜  Desktop time bash s.sh
/System/Volumes/Data/sgoinfre/Perso/llord
bash s.sh  0.31s user 0.60s system 8% cpu 11.053 total
```

I also retried the first standalone commmand and I got this ( it didn't find anything ):

```
➜  Desktop time find / -maxdepth 12 -path "$HOME" -prune -o -path '*/sgoinfre/*' -type d -name "$USER" -print -quit 2>/dev/null | perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/"
find / -maxdepth 12 -path "$HOME" -prune -o -path '*/sgoinfre/*' -type d -nam  4.69s user 15.08s system 75% cpu 26.261 total
perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/"  0.00s user 0.00s system 0% cpu 26.261 total
```
The second standalone command both works and is very fast! :

```
➜  Desktop time (find /sgoinfre /System/Volumes/Data/sgoinfre -maxdepth 6 -type d -name "$USER" -print -quit 2>/dev/null | perl -ne "print \"\$1\n\" if /^(.*?\/sgoinfre\/.*?\/$USER)(?:\/|\$)/")
/System/Volumes/Data/sgoinfre/Perso/llord
( find /sgoinfre /System/Volumes/Data/sgoinfre -maxdepth 6 -type d -name    2)  0.00s user 0.02s system 14% cpu 0.201 total
```

**Comment by itislu** [comment]
@loyc12 Thank you very much, fantastic report!

> I Tried the script and it works!!! I only had to add _bfs after the name of the function called ( you forgot to add it )

Aah sorry that I missed that!
I see that the total time was 11.053 total. I never fully understood the output of time tbh. 
For you the user, do you see the result after 0.6 or 11 seconds?

> I also retried the first standalone commmand and I got this ( it didn't find anything ):

That's quite weird that it doesn't find anything at all. I can't explain it.

> The second standalone command both works and is very fast! :

That's great news!

**Comment by loyc12** [comment]
Turns out we need to use the total values ( so 11.053s, 26.261s and 0.201s ), so that’s my bad. Still, the last command is by far the fastest, tho not necessarily fool-proof, as its a pretrim if I understand well

**Comment by itislu** [comment]
Fixed with commit 937a3e7 in release [v.1.8.0](https://github.com/itislu/42free/releases/tag/v1.8.0)!
